File: src/main/java/com/gmail/borlandlp/minigamesdtools/activepoints/ActivePointController.java
package com.gmail.borlandlp.minigamesdtools.activepoints;

import com.gmail.borlandlp.minigamesdtools.APIComponent;
import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.activepoints.behaviors.Behavior;
import com.gmail.borlandlp.minigamesdtools.config.ConfigEntity;
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath;
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.ArrayList;
import java.util.List;

public class ActivePointController implements APIComponent, ActivePointsAPI {
    private List<ActivePoint> activatedPoints = new ArrayList<>();
    private List<ActivePoint> allPoints = new ArrayList<>();
    private StaticPointsCache staticPointsCache = new StaticPointsCache();
    private Listener listener;
    private BukkitTask task;

    @Override
    public void onLoad() {
        this.listener = new ActivePointsListener(this);
        MinigamesDTools.getInstance().getServer().getPluginManager().registerEvents(this.listener, MinigamesDTools.getInstance());

        final ActivePointController task = this;
        this.task = new BukkitRunnable() {
            public void run() {
                task.update();
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 0, 20);

        // load default activepoints
        List<ConfigEntity> poolContents = MinigamesDTools.getInstance().getConfigProvider().getPoolContents(ConfigPath.ACTIVE_POINT);
        for(ConfigEntity configEntity : poolContents) {
            Debug.print(Debug.LEVEL.NOTICE,"[ActivePointController] load activePoint " + configEntity.getID());
            ActivePoint activePoint = null;
            try {
                activePoint = MinigamesDTools.getInstance().getActivePointsCreatorHub().createActivePoint(configEntity.getID(), new DataProvider());
            } catch (Exception e) {
                e.printStackTrace();
            }

            if(activePoint != null) {
                this.registerPoint(activePoint);
            } else {
                Debug.print(Debug.LEVEL.WARNING,"[ActivePointController] fail on load activePoint " + configEntity.getID());
            }
        }
    }

    @Override
    public void onUnload() {
        this.task.cancel();
        HandlerList.unregisterAll(this.listener);
    }

    public ActivePoint searchPointByID(String ID) {
        for (ActivePoint activePoint : this.getAllPoints()) {
            if(activePoint.getName().equals(ID)) return activePoint;
        }

        return null;
    }

    @Override
    public void registerPoint(ActivePoint point) {
        point.setActivePointController(this);
        this.allPoints.add(point);
        Debug.print(Debug.LEVEL.NOTICE,"[ActivePointController] register activePoint: " + point.getName() + " for controller.");
    }

    @Override
    public void unregisterPoint(ActivePoint point) throws Exception {
        if(this.activatedPoints.contains(point)) {
            throw new Exception("The active point must be deactivated before calling this function.");
        }

        this.allPoints.remove(point);
        point.setActivePointController(null);
        Debug.print(Debug.LEVEL.NOTICE,"[ActivePointController] unregister activePoint: " + point.getName() + " for controller.");
    }

    @Override
    public void activatePoint(ActivePoint point) {
        point.spawn();
        activatedPoints.add(point);
    }

    @Override
    public void deactivatePoint(ActivePoint point) {
        if(point != null && activatedPoints.contains(point)) {
            point.despawn();
            activatedPoints.remove(point);
        }
    }

    private void update() {
        for (ActivePoint activePoint : this.activatedPoints) {
            for (Behavior behavior : activePoint.getBehaviors()) {
                behavior.tick();
            }
        }
    }

    @Override
    public StaticPointsCache getStaticPointsCache() {
        return staticPointsCache;
    }

    @Override
    public List<ActivePoint> getAllActivatedPoints() {
        return new ArrayList<>(this.activatedPoints);
    }

    @Override
    public List<ActivePoint> getAllPoints() {
        return new ArrayList<>(this.allPoints);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/activepoints/ActivePointController.kt
package com.gmail.borlandlp.minigamesdtools.activepoints

import com.gmail.borlandlp.minigamesdtools.APIComponent
import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

class ActivePointController : APIComponent, ActivePointsAPI {
    override val activatedPoints: MutableList<ActivePoint> = mutableListOf()
    override val allPoints: Queue<ActivePoint> = ConcurrentLinkedQueue()
    override val staticPointsCache = StaticPointsCache()
    private var listener: Listener? = null
    private var task: BukkitTask? = null

    override fun onLoad() {
        listener = ActivePointsListener(this).apply {
            instance!!.server.pluginManager.registerEvents(listener, instance)
        }
        val task = this
        this.task = object : BukkitRunnable() {
            override fun run() {
                task.update()
            }
        }.runTaskTimer(instance, 0, 20)
        // load default activepoints
        val poolContents =
            instance!!.configProvider!!.getPoolContents(ConfigPath.ACTIVE_POINT)
        for (configEntity in poolContents) {
            print(
                Debug.LEVEL.NOTICE,
                "[ActivePointController] load activePoint " + configEntity.id
            )
            var activePoint: ActivePoint? = null
            try {
                activePoint = instance!!.activePointsCreatorHub!!.createActivePoint(configEntity.id, DataProvider())
            } catch (e: Exception) {
                e.printStackTrace()
            }
            if (activePoint != null) {
                registerPoint(activePoint)
            } else {
                print(
                    Debug.LEVEL.WARNING,
                    "[ActivePointController] fail on load activePoint " + configEntity.id
                )
            }
        }
    }

    override fun onUnload() {
        task!!.cancel()
        HandlerList.unregisterAll(listener)
    }

    override fun searchPointByID(ID: String): ActivePoint? {
        for (activePoint in allPoints) {
            if (activePoint.name == ID) return activePoint
        }
        return null
    }

    override fun registerPoint(point: ActivePoint) {
        point.activePointController = this
        allPoints.add(point)
        print(
            Debug.LEVEL.NOTICE,
            "[ActivePointController] register activePoint: " + point.name + " for controller."
        )
    }

    @Throws(Exception::class)
    override fun unregisterPoint(point: ActivePoint) {
        if (activatedPoints.contains(point)) {
            throw Exception("The active point must be deactivated before calling this function.")
        }
        allPoints.remove(point)
        point.activePointController = null
        print(
            Debug.LEVEL.NOTICE,
            "[ActivePointController] unregister activePoint: " + point.name + " for controller."
        )
    }

    override fun activatePoint(point: ActivePoint) {
        point.spawn()
        activatedPoints.add(point)
    }

    override fun deactivatePoint(point: ActivePoint) {
        if (activatedPoints.contains(point)) {
            point.despawn()
            activatedPoints.remove(point)
        }
    }

    private fun update() {
        for (activePoint in activatedPoints) {
            for (behavior in activePoint.behaviors) {
                behavior.tick()
            }
        }
    }
}